### ISSUE TITLE
Allow listening on unix sockets

### ISSUE THREAD

**Issue by CKing** [nats-io/gnatsd#320]
If I use a service only on localhost, i don't like blocking a port and prefer using a unix socket instead.
Is it possible to add unix sockets for listening on connections?

**Comment by derekcollison** [comment]
While possible, there is no performance advantage on modern linux kernels to really warrant the additional code. You can specify which network you would like to bind the listen port.

**Comment by kanaka** [comment]
One example where this would be useful is in a docker-compose context where we don't have fully connectivity between the client containers but can volume mount a shared unix domain socket to send and receive messages from and the NATS server.

**Comment by mutech** [comment]
Reiterating on this: While performance might not make a big difference, the ability to determine the UID and PID of the client does.

It allows a local server to authenticate the client process by IP. It can also use the PID (given appropriate permissions) to perform additional checks on the process (was it started by a parent -> is it running as some systemd unit) and more.

This would allow a system-nats instance to authorize system service securely without the need to deploy certs or other secrets. It can even act as credential broker on this basis and integrate services on the system and the system as a whole into a local network via nats.

This would greatly help on-prem and edge deployments to integrate into cloud and remote systems. It would be much easier to build such systems, provide more observability, auditing and security.

The reason why this is so big is because systems (OS) have well defined static security boundaries (UID & FS perms). When system boundaries are broken, local nats deployments have little protection (from local root). Consequently, trusting this local boundary is not much of a loss. But when you already have to trust this boundary, you gain all the benefits of already established permission models.

Using a TCP connection instead means that each client has to establish a trust relationship with NATS individually, and usually that means having to provide some secret. This secret does not only depend on UID, but also on the protection of the secret on the file system or the communication channel through which the client obtains it. This is less of a boundary than the possession of the UID (a file might be accessible to other UIDs) and even less than a process with a UID having a specific PPID (or other observable properties, etc. /proc/PID/**.

I can write a wrapper around a nats server (e.g. embedded nats), accept connection on a domain socket and either use this socket in place of a TCP socket (are there any problems with this?), but this would have some serious problems: First I have to do that right. I'm confident that I can, but I'm not a security expert. That's good enough for me, but for many users, that's an argument not to use such a mechanism with serious security implications. I had to maintain the wrapper and update it myself. Thus I loose automatic updates - on all machines. Lastly I need to deploy a custom solution instead of just choosing a packet, even more administrative overhead.

Having such a system-nats as a (soon to be) standard component, would completely change the game of sysops in homelabs and small business, both of which are not cloud friendly. It could become a core component for site networks, because it not only helps to improve security, it would also inject all the flexibility and power of nats into an area that has nothing comparable to offer. Nats is uniquely positioned to fulfill this role, because of its unique feature set.

Just a few examples:

- Can I run a scheduled backup that would take a service down? Today, this is a hard problem. Generally, the backup just runs, disrupting clients. This had to be scheduled if disruptions are a problem. This creates service windows. These windows cannot be changed easily, depending on who relies on that. This creates geographic boundaries (time zones). With nats, this can be easily solved. With system nats, servers in a local network become trivially visible and with system-nats the effort to employ nats in backup scripts and service clients becomes trivial. No need to setup secure communication channeles between client and backup, no need for scheduling. A client simply needs a priority assigned to its local credentials (resulting in the permission to block a backup for certain period) or some similar mechanism reacting to a client starting a process (nats message). If integrating nats becomes a frictionless process (call the nats cli), then it becomes really attractive. If you have to authorize and authenticate each and every client, it's a nightmare.
- If you have to deploy secrets for services (non-local scope), you can easily correlate that boot cycles or service life cycles and only serve the first request for a secret, alert or dead-man-switch on subsequent requests. As such, secrets never need to touch the file system and their provisioning is completely auditable.
- To implement centralized logging, you typically install scraping agents collecting metrics and logs. Each host needs that extra setup for then logging and metrics. For similar use cases, similar setups are needed. The scraping/deployment has to be proactive and comes at some cost. If that information would be channeled through nats, it could become on-demand services only active when in use. Access to these data streams can be modeled with fine granularity. Alerts often depend on the data to walk through the processing cycle of logs and metrics, ranging from seconds to minutes. Nats could be so much more reactive.

In short, having local authentication through unix domain sockets would open up a whole market (albeit possibly a small one) to nats that is currently unavailable, due to the administrative burden of authentication or security concerns with unauthorized clients. Considering the increasing importance of data sovereignty, that market might not be quite so small though.

I opened a discussion about this here: https://github.com/nats-io/nats-server/discussions/7677
